data/src/main/java/org/apache/iceberg/MetricsAppender.java

```diff
@@ -0,0 +1,839 @@
+      return this;
+    }
+
+    public Builder useIntForDateType() {
+      dateTypeRepresentation = DateTypeRepresentation.INT;
+      return this;
+    }
+
+    public Builder useLocalDateForDateType() {
+      dateTypeRepresentation = DateTypeRepresentation.LOCAL_DATE;
+      return this;
+    }
+
+    public Builder useLongForTimeType() {
+      timeTypeRepresentation = TimeTypeRepresentation.LONG;
+      return this;
+    }
+
+    public Builder useLocalTimeForTimeType() {
+      timeTypeRepresentation = TimeTypeRepresentation.LOCAL_TIME;
+      return this;
+    }
+
+    public Builder useLongForTimestampType() {
+      timestampTypeRepresentation = TimestampTypeRepresentation.LONG;
+      return this;
+    }
+
+    public Builder useDateTimeForTimestampType() {
+      timestampTypeRepresentation = TimestampTypeRepresentation.OFFSET_DATETIME;
+      return this;
+    }
+
+    public Builder useLocalDateTimeForTimestampType() {
+      timestampTypeRepresentation = TimestampTypeRepresentation.LOCAL_DATETIME;
+      return this;
+    }
+
+
+    public MetricsAppender<D> build() {
+      // TODO check iif not defaults??
+      return new MetricsAppender<>(appender, schema,
+          new MetricsAppenderConfiguration(recordRepresentation,
+              fixedTypeRepresentation,
+              dateTypeRepresentation,
+              timeTypeRepresentation,
+              timestampTypeRepresentation));
+    }
+  }
+
+  private static class MetricsCollectorMap<K, V> extends MetricsCollectorBase<Map<K, V>> {
+    private final MetricsCollector collectorKey;
+    private final MetricsCollector collectorValue;
+
+    MetricsCollectorMap(MetricsCollector<K> collectorKey, MetricsCollector<V> collectorValue) {
+      this.collectorKey = collectorKey;
+      this.collectorValue = collectorValue;
+    }
+
+    @Override
+    public void add(Map<K, V> map) {
+      if (map != null) {
+        for (K k : map.keySet()) {
+          collectorKey.add(k);
+          collectorValue.add(map.get(k));
+        }
+      } else {
+        collectorKey.add(null);
+        collectorValue.add(null);
+      }
+    }
+
+    @Override
+    public Stream<FieldMetrics> getMetrics() {
+      return Stream.of(collectorKey, collectorValue).flatMap(MetricsCollector::getMetrics);
+    }
+  }
+
+  private static class MetricsCollectorList<L> extends MetricsCollectorBase<List<L>> {
+
+    private final MetricsCollector collector;
+
+    MetricsCollectorList(MetricsCollector<L> collector) {
+      this.collector = collector;
+    }
+
+    @Override
+    public void add(List<L> list) {
+      if (list != null) {
+        for (L e : list) {
+          collector.add(e);
+        }
+      } else {
+        collector.add(null);
+      }
+    }
+
+    @Override
+    public Stream<FieldMetrics> getMetrics() {
+      return collector.getMetrics();
+    }
+  }
+
+  private static class MetricsCollectorGenericRecord extends MetricsCollectorBase<GenericData.Record> {
```

Comment:
In most cases, we use `IndexedRecord` instead of `GenericData.Record`. That's the superclass of both `GenericData.Record` and `SpecificRecord` so you get support for Avro generics and Avro specifics.